feat: add support for nested objects via `prop` object

JWThewes opened the pull request:
Add support for nested objects. This allows to access properties of nested objects via "." notation.

imballinst commented on src/__stories__/00-Uncontrolled/08-NestedObject.tsx:
```diff
@@ -0,0 +1,37 @@
+import { TableHeader } from '../../components/TableHeader';
+import { TableBody } from '../../components/TableBody';
+
+// @@@SNIPSTART CustomRowStyle
```
Could we update this `CustomRowStyle` into `NestedObject`? This part is used to synchronize the snippets in https://imballinst.github.io/react-bs-datatable/examples/, later on I can create a new page consisting this one.

imballinst reviewed (changes requested):
Thanks @JWThewes for implementing this, I think this is a good feature.

I wonder if we could enhance `TableColumnType['prop']` so that it can derive nested prop, something like this? https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgCoQM5gPICMBWECYyA3gFDLJgD2ADgDIQBuEANgFzJZSgDmlZAAt2bGlwpUqITJAAmXHv0FUA7jShsFZFVICuIYKygYIisLxACpUuvy4AlIhrkAeJVYA0yOCACeAHy6AL6CoaHkYH50KADiEGAAgmxsANIQfhiuqAHIALxoyBAAHpAgchjINAREJAD8ZMgA2qnIoMgA1hk0MGgAuhyCAAYAJKQAosUIbHpyEK6p3hh+ALa4NGwBwWMARDvIAD7IQwB0Y-FJKemZ2S19W0PBQ+TBTV1+Pf3IXDLG5OQIGggLDIYpcC7JNIZLLoLB4QjEXIFHYiFI0E7qTRyE52Kw7IA

So we don't have to manually inject `rocket.name` and `rocket.company` to `StoryColumnType`.

JWThewes commented:
Quick update:
I fixed the stuff you mentioned (Formatting and copy&past error).
I also added a first version of NestedKey support. Therefore I changed the generic of TableColumnType to <T extends object> hope that is ok?
One last bit is missing. In my case rocket is optional on the data. It all works but inside the IDE you get an error for optional types.
As of now I have no idea how to support optional types with NestedKeyOf.

If you have any idea happy to see what you can do here...

imballinst commented:
> I fixed the stuff you mentioned (Formatting and copy&past error).
> I also added a first version of NestedKey support. Therefore I changed the generic of TableColumnType to hope that is ok?

Thanks! Appreciate it. And yeah, it _seems_ fine to me.

> One last bit is missing. In my case rocket is optional on the data. It all works but inside the IDE you get an error for optional types.
> As of now I have no idea how to support optional types with NestedKeyOf.

hmmm, that's a good point. Let me check on my end and see if I can help with it.

imballinst commented:
I have tried to fix the thing, mostly around 3 parts:

1. Add the `engine` field to the story data
2. Make `DatatableWrapper` generic type to also `extends object`
3. and somewhat follow the original `GetNestedKeys` utility type

Let me know if you have concerns about these updates @JWThewes, thanks! If not, then I think I can merge this and release it sometimes tomorrow my morning.

JWThewes commented:
Looks good. Thanks for helping out!
Looking forward to that release 👌

imballinst commented:
This PR has now been released: 🎉

- https://github.com/imballinst/react-bs-datatable/releases/tag/v3.14.0
- https://www.npmjs.com/package/react-bs-datatable/v/3.14.0

Let me know if you have issues with the latest build. Thanks!

JWThewes commented:
It's working so far. Only thing missing is sorting and filter support for nested properties. Totally forgot about these. Will have to take a look to make it working...

imballinst commented:
......that's a good point. I guess we need to expand the Storybook story a little bit and add tests for the sort/filter. Should be relatively straightforward, though, we can just replace `row[prop]` with `getProp(row, prop)` where `getProp` is our `.reduce()` thingy.

JWThewes commented:
Just opened the PR https://github.com/imballinst/react-bs-datatable/pull/206 to add sorting and filtering support. Happy to get your feedback and see it merged and released :)
